### ISSUE TITLE
Only show relevant omics type results for KEGG search

### ISSUE THREAD

**Issue by naglepuff** [microbiomedata/nmdc-server#1176]
The sample level is not filtering as expected.
Can you make it so that results are only shown when that workflow produces a result? (ie metabolomics & proteomics should not be shown under samples for a search for KEGG 00005 b/c none are shown in the study summary)

_Originally posted by @aclum in https://github.com/microbiomedata/nmdc-server/issues/1072#issuecomment-1974210524_

**Comment by naglepuff** [comment]
@aclum would this be a special case for KEGG terms? For example, this search: https://data-dev.microbiomedata.org/?q=ChQIABACGAIiDCJNZXRhZ2Vub21lIg== narrows down omics type, and that is reflected by the chips under the studies, but the samples still show all of the results across all workflow types.

Based on our conversation today, my understanding of the desired behavior is this:

1) I search for KEGG term `ABC`
2) I scroll down to samples, and `sample_1` matches the search. It might come from the result of a metaG or metaP workflow
3) If the result matches from a metaG workflow, I should only see the `metagenome` tab under the sample. Vice versa if the link is through a metaP workflow
4) Under the tab that shows up (either metaG or metaP), I should see all workflows and data objects of that type

Please let me know about the question at the top of the comment, and correct any misunderstandings I might have about the expected behvior.

**Comment by aclum** [comment]
after a conversation with @emileyfadrosh we decided for consistency between studies and samples if an omics type filter is applied it should remove the other workflows from the sample view. So no, this is not a special case for KEGG terms. 
The rest of the desired behavior you describe is correct.

**Comment by naglepuff** [comment]
> It seems that the desired behavior is supported [directly by SQLAlchemy](https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html#routing-explicit-joins-statements-into-eagerly-loaded-collections) by eager-loading the associated records when querying some table (in this case, loading the `OmicsProcessing` records at the same time as the `Biosample`s.
> 
> I'll start by trying to use this approach. Since these queries can be quite complex at times it might not be as straightforward as the example indicates.

Turns out that this approach is more complicated than I initially thought.

The SQL result of a query using the eager loading has multiple rows per biosample, so serialization using our current offset/limit pagination is negatively impacted. This is reflected by the data portal when implementing this techinique:

![image](https://github.com/microbiomedata/nmdc-server/assets/7085625/846771ac-bcf7-432d-8217-a8fa5e489832)
Here, the page size is set to 15, but only 5 biosample records are listed. Each of these biosamples has 3 omics types.

![image](https://github.com/microbiomedata/nmdc-server/assets/7085625/a1c21bb4-2161-477e-9e6b-0ac47f63d566)
Here, the page size is set to 5, but only 2 biosamples are shown. All 3 omics types for the first sample are listed, but only 2 omics types for *gold:Gb0126426* are listed (compared to 3 in the example above).

Besides eager loading, there are some additional approaches that might work at either the pagination level or the serialization level. However, these approaches would either be more complex or introduce poor performance, as filtering omics_processing records would have to be done at the SQL level, and then additional application logic would have to be applied to either exclude certain records or mark them as not actual results. In either case, there doesn't appear to be a straightforward solution at this time, so I will have to continue investigating the best path forward and reach out to those with some more expertise.

**Comment by ssarrafan** [comment]
Appears to be active. Moving to next sprint. @aclum @naglepuff

**Comment by ssarrafan** [comment]
@naglepuff said this one needs to continue to the next sprint at the infra sync meeting today.
FYI @aclum

**Comment by naglepuff** [comment]
I've opened up #1213 as a (draft as of 4/18) pull request. I believe I've implemented this feature (initial testing looks promising), but I have some concerns with respect to performance (and to a lesser extent, correctness). I think more time testing this would be a good thing, so I'd prefer not to rush this through to dev, especially considering the upcoming 4/23 release.

This should be moved to next sprint again.
